File: cpp/split_pdf/split_pdf.cpp
// split_pdf -- C++ port of examples\Vb6\split_pdf\split_pdf.bas
// Opens one import file, keeps it open across CloseFile() via
// SetUseGlobalImpFiles, and writes each page into its own PDF.
#include "apputil.h"
#include <cstdio>

static SI32 PDF_CALL PDFError(void* Data, SI32 ErrCode, const char* ErrMessage, SI32 ErrType){
    if(ErrMessage) printf("%s\n", ErrMessage);
    return 0;
}

int main(){
    ChdirToExe();
    PPDF pdf = pdfNewPDF();
    pdfSetOnErrorProc(pdf, 0, PDFError);

    pdfSetImportFlags(pdf, ifImportAll | ifImportAsPage);  // avoid conversion of pages to templates
    pdfSetImportFlags2(pdf, if2UseProxy);                  // reduces the memory usage

    const char* inFile = "license.pdf";
    if(pdfOpenImportFileA(pdf, inFile, ptOpen, "") < 0){
        pdfDeletePDF(pdf);
        return 0;
    }

    // Keeps the open import file from being closed when CloseFile() is called.
    pdfSetUseGlobalImpFiles(pdf, 1);

    const char* outDir = "out";
    CreateDirectoryA(outDir, 0);

    SI32 count = pdfGetInPageCount(pdf);
    for(SI32 i = 1; i <= count; i++){
        char outPath[512];
        sprintf(outPath, "%s\\page%04d.pdf", outDir, (int)i);
        pdfCreateNewPDFA(pdf, outPath);
        pdfAppend(pdf);
        pdfImportPageEx(pdf, i, 1.0, 1.0);
        pdfEndPage(pdf);
        pdfCloseFile(pdf);
    }

    // Always set the property back to false when finished.
    pdfSetUseGlobalImpFiles(pdf, 0);

    printf("Pages written to: %s\n", outDir);
    pdfDeletePDF(pdf);
    return 0;
}

File: csharp/split_pdf/split_pdf.cs
//  split_pdf -- C# port of examples\Vb6\split_pdf\split_pdf.bas
//  Opens one import file, keeps it open across CloseFile() via
//  SetUseGlobalImpFiles, and writes each page into its own PDF.
using System;
using System.IO;
using LumasPdfSdk;

class SplitPdf
{
    static int PDFError(IntPtr Data, int ErrCode, string ErrMessage, int ErrType)
    {
        Console.WriteLine(ErrMessage);
        return 0;
    }

    static TErrorProc _errCb = PDFError;

    static void Main()
    {
        IntPtr pdf = LumasPdf.pdfNewPDF();
        LumasPdf.pdfSetOnErrorProc(pdf, IntPtr.Zero, _errCb);

        LumasPdf.pdfSetImportFlags(pdf, LumasPdfConsts.ifImportAll | LumasPdfConsts.ifImportAsPage); // avoid conversion of pages to templates
        LumasPdf.pdfSetImportFlags2(pdf, LumasPdfConsts.if2UseProxy);                                // reduces the memory usage

        // Original: '../../../license.pdf' -> resolved to the bundled import fixture.
        string inFile = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "license.pdf");
        if (LumasPdf.pdfOpenImportFileW(pdf, inFile, (int)LumasPdfConsts.ptOpen, "") < 0)
        {
            LumasPdf.pdfDeletePDF(pdf);
            return;
        }

        // Keeps the open import file from being closed when CloseFile() is called.
        LumasPdf.pdfSetUseGlobalImpFiles(pdf, true);

        string outDir = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "out");
        Directory.CreateDirectory(outDir);

        int count = LumasPdf.pdfGetInPageCount(pdf);
        for (int i = 1; i <= count; i++)
        {
            string outPath = Path.Combine(outDir, "page" + i.ToString("0000") + ".pdf");
            LumasPdf.pdfCreateNewPDFW(pdf, outPath);
            LumasPdf.pdfAppend(pdf);
            LumasPdf.pdfImportPageEx(pdf, (uint)i, 1.0, 1.0);
            LumasPdf.pdfEndPage(pdf);
            LumasPdf.pdfCloseFile(pdf);
        }

        // Always set the property back to false when finished.
        LumasPdf.pdfSetUseGlobalImpFiles(pdf, false);

        Console.WriteLine("Pages written to: " + outDir);
        LumasPdf.pdfDeletePDF(pdf);
    }
}
